### ISSUE TITLE
OutOfMemoryErrors after updating to 4.3

### ISSUE THREAD

**Issue by kschults** [robolectric/robolectric#5131]
### Description

After updating from 3.x to 4.3 (using the auto-migration), most of the tests in my project are fine, but I'm getting OutOfMemoryError on many (but not all) tests when running the full test suite. When running failing tests individually, there's no problem.

### Steps to Reproduce

My test suite has ~1200 tests. The errors consistently start happening ~700 tests in, and there are ~80 failures over the final 500 tests. It seems that they are mostly in test cases where I am creating an Activity and/or Fragment. I have tried using `ActivityScenario` and `FragmentScenario`, as well as reverting to `Robolectric.buildActivity`, but with the same results.

### Robolectric & Android Version

Robolectric 4.3
Android compileSdkVersion 28
Android targetSdkVersion 27

**Comment by byronroberts** [comment]
Having the exact same issue after we have just completed a migration from 3.8 to 4.3. We migrated ours manually rather than use the migration tool. Also migrated our project to Androidx and followed the migration guide to change to the recommended Androidx test tools.

**Comment by bhidalgo1999** [comment]
Out of curiosity, did you guys update to Grande 5.x? Heap memory was halved from 1GB.

**Comment by byronroberts** [comment]
Yeah I'm currently using Gradle 5.2 and set the following arguments in the gradle.properties file org.gradle.jvmargs=-Xss2m -Xmx2g. Tried changing this to 4g instead of 2g to no avail.

**Comment by Davidaz4** [comment]
Same problem here with robolectric 4.4. Just in case I tried to enable largeHeap on the android manifest with no results. I have Gradle 5.1.1. 

I have lots of errors with Gradle 5.1.1 and robolectric in general, if I stay on 4.0 beta (the version I had before updating) I receive Null Pointer exceptions that make nonsense, or manifest missing on the 4.0 stable library. 

I have to roll back to previous Gradle if I want to solve it.

**Comment by kschults** [comment]
I'm on Gradle 5.2, and running with `org.gradle.jvmargs=-Xmx3072m`. It runs out of memory, and then everything grinds to a halt. 

I'm trying to update because another update in my project requires switching to AndroidX, which means I need to update Robolectric as well.

**Comment by bhidalgo1999** [comment]
Please try `jvmArgs = “-Xmx1g` within the `unitTest` closure in your module level gradle file. The one in gradle properties does not díctate heap allocation for forked tasks! This is just a work around. Either Robolectric has a thicc memory profile, or you have leaky test suites!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by bhidalgo1999** [comment]
Or whatever the syntax is, commenting through my iPhone. Sorry.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by byronroberts** [comment]
@bhidalgo1999 Adding that in does nothing for us. I would go down the route of Robolectric having a leak somewhere as our tests ran fine and in good time while we were using 3.8. 4.3 has completely wiped us out. 

If you use AndroidX, am I right in assuming you have to upgrade Robolectric too? Ideally we would just stay on 3.8 and use the AndroidX libraries at this point.

**Comment by Davidaz4** [comment]
@byronroberts In my case, when I started using androidX I was on 4.0-beta without any problem, so as long as you don't update gradle to 5.x it shouldn't be any issue (you can try with 4.3 and androidX and I suppose it will also work)

Regarding what @bhidalgo1999 said about the jvmArgs, I have been trying to increase the heap on that way on the unitTest but with gradle 5 syntax and it didn't do anything. I even tried to get more data about the error with the _jvmArgs '-XX:+HeapDumpOnOutOfMemoryError'_ flag, in order to see more info but it didn't show anything more than that error.

I will look for any possible leaky test as you say, but I don't think that's the reason. I will still try to force the heap on every possible way, if I find any workaround I will write back.

**Comment by bhidalgo1999** [comment]
Huh. Interesting. Our project uses AndroidX, Robo 4.3, latest gradle plugin and wrapper...and it just happened to be the heap cut that caused our problem (we have ~1000 tests and about 10% use FragmentScenario). I’d say use a tool like visualvm to see what exactly is going on. Sorry fellas, was hoping I could help!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by bhidalgo1999** [comment]
I guess it’s worth noting which way are you guys running your tests? Through a run configuration in the IDE? Using the command line interface? Also do you have an accurate read on how much memory is actually allocated? If you’re providing Jvm Args and they are being ignored, that would be important to note. You might have to pass them in through `JAVA_OPTS` if using the command line client (only 64MB allocated as of Gradle 5.x).

Good luck guys!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by byronroberts** [comment]
@bhidalgo1999 turns out what you mentioned with adding the jvmArgs = “-Xmx1g” directly into each modules build.gradle file worked for us although we have allocated more memory that the 1gb! We initially had other arguments going into that field too which it didn't like and just wanted the memory allocation on it's own. We now have them running as we want.

I think the issue could be that the memory allocation you set in the gradle.properties is either being ignored or overridden somewhere else when running Robolectric which was causing us the issues. We are using AGP 3.4.1 and Gradle 5.4.1

**Comment by bhidalgo1999** [comment]
@byronroberts That's awesome! The confusing part about allocating memory to the daemon is that there are many different ways and each serve their own purpose. I know for a fact that the gradle.properties/org.gradle.jvmargs or whatever allocates memory to the daemon as a whole. The problem is that running a test task is technically a forked process and only gets a chunk of that memory you initially declared. That's why specifying it within the unitTest closure works. Apart from that, running your test task through a AS run config will allocate the memory you specify in your custom VM properties...And finally the CLC, which uses the memory you allocate through the JAVA_OPTS envvar.

Here at my current project, we ended up giving 1GB, but I am actively using tools like VisualVM to find objects that aren't being collected between test cases so that I could thin down our memory usage as a whole which will lead to faster times 🤞. Anyways, glad you got it working and goodluck!

**Comment by kschults** [comment]
@bhidalgo1999 I'm getting the same results running both from the command line and from an IDE configuration. I'll keep trying to experiment with the memory allocation, but that's still just a bandaid on what seems to be a memory leak within Robolectric.

If anyone has suggestions on how to further investigate the leak, I'm happy to look. I'm just not terribly familiar with all the different memory options or how to examine heap dumps.

**Comment by bhidalgo1999** [comment]
@kschults I totally agree. Definitely just a work around.

Like I mentioned above, I use VisualVM to profile unit test memory usage. Heap dumps are kinda intimidating at first, but I suggest filtering out by your package name and see what objects have the highest retained amounts of memory.

I'm keeping my eye on this issue to see when one of the active contributors of 4.0 weigh in. If there are efforts to thin down Robo's memory usage, I ain't complainin' :).

**Comment by Davidaz4** [comment]
@bhidalgo1999 are you using VisualVM as a Intellij plugin or as an external tool? I am now downloading the plugin, so I could see hopefully where I could be leaking memory and try to solve it. 

If not I will download the tool itself and launch the gradle command

Thank you for recommending the tool, it will be really useful!!

**Comment by Davidaz4** [comment]
Hello again. With the VisualVM tool, and the @bhidalgo1999 advices I managed to do a full test run. My problem was that, as @bhidalgo1999 said, I had memory leaks on my tests. It turned out I never freed the ActivityController's activities, so they were keep in memory. I didn't even needed the larger heap, it was all my problem.

Maybe it didn't happened on previous versions as Gradle's heap was bigger on previous versions.

To @kschults , and maybe @byronroberts , if you happen to make the same mistake as myself, I solved it with an `@After` task on every Test Class that used an ActivityController, where I destroyed the created activity (my mistake, thinking it was automatic from a long time), calling `activityController.pause().stop().destroy()`

Thanks for the advices!!

**Comment by markchristopherng** [comment]
@bhidalgo1999 are you able to share your build.gradle file so we see how it was configured?

**Comment by bhidalgo1999** [comment]
@markchristopherng Unfortunely, this is all I can share:

`
android {
     ...
     testOptions {
          ...
          unitTests.all {
               ...
              jvmArgs "-Xmx1g" //(or whatever amount of memory your suites need).
               ...
          }
          ...
     }
     ...
}
`

This is a solution for one module. If you have a multimodule project, chances are this block is shareable. You could try to fit this in your project level gradle file under subprojects (?). But this is how you'd allocate more memory to your unit test subtask initialized through the CLI for a monolithic project.

To be clear, @kschults is right (in my opinion), your tests should be able to run under the new Gradle 5 standard of 512MB. This is still an ongoing effort on my part at my company, but I am getting very close through optimizations to our test suite.

If anyone thinks it's worth consulting me, feel free to send me a PM or email @ bhidalgo1999@gmail.com!

**Comment by markchristopherng** [comment]
Thanks that's a great help

**Comment by kschults** [comment]
@bhidalgo1999 Thanks for the suggestions. Thanks also to @Davidaz4 for the comment about `ActivityController` needing to be `destroy`ed after each task.

Between cleaning up my controller usage and increasing the memory specifically for the test tasks, I was able to get my suite to complete.

That said, I'm fairly sure there's at least a couple leaks in Robolectric 4. For one, when using `FragmentScenario`, it internally creates an `ActvityScenario`, which gets added to the static`ActivityLifecycleMonitorRegistry` and never removed because there's no way to call `destroy` on the internal `ActivityScenario`

Beyond that, even after removing `FragmentScenario` from my tests, I was still running out of memory, which seems to indicate at least one other significant leak.

**Comment by bhidalgo1999** [comment]
@kschults As for the Scenario classes, I'm not sure if that has anything to do with Robo. Aren't they part of the Fragment Testing library? Also out of curiosity, anyway you can do something like:

```
scenario.onFragment {
     it.requireActivity().finish()
}
```
(Lol)

I never tried it myself, but I wonder what can of worms that would open.

**Comment by kschults** [comment]
Ah, you're right, that is part of AndroidX, not part of Robo.

Based on a quick inspection, I can't see anything inside `ActivityScenario` that would detect the activity being finished and clean itself up. The lifecycle monitor appears to only be used to detect when a state change requested through the scenario has completed.

**Comment by frank-tan** [comment]
I had this problem in the last couple of days. The information in this thread helped massively. To make it easier for other people, I will share my solution and experience here.

After upgrading to Robolectric 4.3 and `Gradle 5.1.1`, I started to have hanging build in commandline using `./gradlew testDebugUnitTest` and in the build server. But building in Android Studio was fine. I suspect Android Studio is using its own Gradle config. (If you know where its config is, please let me know 😀)

After monitoring the memory consumption using VisualVM (https://visualvm.github.io/), I found Gradle 5.1.1 has a much smaller heap size (500MB) as opposed to Gradle 4's 4GB heap size.

![Gradle 4 - 4GB heap size](https://user-images.githubusercontent.com/3487861/60379314-4c63bf00-9a74-11e9-94b9-5ee158b4b02a.png)

![Gradle 5 - 500 MB heap size](https://user-images.githubusercontent.com/3487861/60379317-5eddf880-9a74-11e9-99ce-20d31138b47c.png)

Somehow my application needs more than that to run the test. The simplest solution is to increase the build heap size by adding the custom heap size in `app/build.gradle`. I set it to 2GB here

```
android {
  testOptions {
    unitTests.all {
      jvmArgs '-Xmx2g'
    }
  }
}
```

This should solve the problem. But if you need that much memory for running unit tests, most likely you are leaking memory in the code. Upon further investigation with `VisualVM`'s heapdump taken from the hanging build, I fixed the memory leak and 500MB is actually quite enough.

In summary, Robolectric is not the root cause of the OutOfMemoryError. The root cause is a combination of memory leak in my code and Gradle 5 upgrade. To work around the issue quickly, change the memory setting for the unit test in Gradle. To fix it more elegantly or correctly, try to find and fix your memory leak. 500MB should be enough for any unit test.

**Comment by kingargyle** [comment]
Another suggestion, is to setup your gradle build in your root build.gradle to Fork the JVM and run tests in parallel.   There are still issues with Robolectric and especially large test suites that it doesn't free up memory.  Increasing heap will solve it for a bit but eventually you hit diminishing returns.   We use the following in our builds:

`subprojects {
  tasks.withType(Test) {
    maxParallelForks = 2
    forkEvery = 80
    maxHeapSize = "2048m"
    minHeapSize = "1024m"
  }
}`

This will set the max heap to 2048, and a min heap of 1024, it will fork the build every 80 or so test classes, and run tests in parallel.

**Comment by hoisie** [comment]
Closing this as 4.3 is obsolete now.
